lib/srv/db/spanner/listener.go

```diff
@@ -0,0 +1,68 @@
+/*
+ * Teleport
+ * Copyright (C) 2024  Gravitational, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package spanner
+
+import (
+	"context"
+	"io"
+	"net"
+	"sync/atomic"
+)
+
+// fakeListener pretends to be a listener, but all it does is Accept() a single
+// connection and then block until its context is done or Close is called.
+type fakeListener struct {
+	conn         net.Conn
+	accepted     atomic.Bool
+	closeContext context.Context
+	cancel       context.CancelFunc
+	acceptHook   func()
+}
+
+func newFakeListener(ctx context.Context, c net.Conn, hook func()) *fakeListener {
+	ctx, cancel := context.WithCancel(ctx)
+	return &fakeListener{
+		conn:         c,
+		closeContext: ctx,
+		cancel:       cancel,
+		acceptHook:   hook,
+	}
+}
+
+func (l *fakeListener) Accept() (net.Conn, error) {
+	if l.accepted.Swap(true) {
+		<-l.Done()
+		return nil, io.EOF
+	}
+	l.acceptHook()
```

Comment:
You're right that it essentially measures the time it takes to setup the server for connections. It looks like most of our engines do as you say, whereas the http based ones like elasticsearch, opensearch, dynamo, snowflake do not, so that wasn't clear to me.

Would you prefer this to measure the time it takes to check access for the first RPC, emit our `db.session.start` event and dial GCP?

Reply:
i don't think the connection setup metrics description makes sense for HTTP and GRPC engines, as these engines do NOT setup db connections before any requests are received. Personally I would rather introduce new metrics if needed, if we really want to measure http/GRPC dialling or round trip time.

That said, do we count `common.GetMessagesFromServerMetric` or `common.GetMessagesFromClientMetric` in spanner?

Reply:
we can add those metrics in the interceptors. Although for the server streaming RPCs, there can be multiple messages received from the server, so we might count each of them as a message

Reply:
actually this is probably better handled in a `grpc.StatsHandler` which we can install on both the server and the internal client that we use to send RPCs to GCP.

i.e.:
`grpc.StatsHandler` (server option)
`grpc.WithStatsHandler` (client dial option)

Reply:
> Number of messages (packets) received from the DB client.

In context of how we implemented these metrics, the "packets" are at the application layer as in "mysql protocol packet", not a TCP packet, so I think an RPC is equivalent.

Also, it looks like we wrap `engine.SendError` to report `teleport_db_errors_total`. Which made me think of your other suggestion to maybe return handler errors from the engine's `HandleConnection` function - but I think I can do that in the interceptor instead which will increment that metric properly for all RPC handlers that send a synthetic error from Teleport.

Reply:
Hmm, looking at how the mongo engine does it, and since I shutdown the grpc server if any RPC is denied access, I could instead save the error which triggers a shutdown and then return that from the handler so it gets reported as just one synthetic error message sent to the client, even though potentially multiple RPCs got that error message.

Reply:
Ah ! I think it's ok if `engine.SendError` does not end up doing anything for spanner but it would be good to at least for HandleConnection to return a copy of the error for reporting purpose.

Reply:
Because of the potential for many RPCs to be in flight and get errors at the same time, and because SendError only gets called for the wrapper on one error, I instead increment that error metric manually. See: [ca49e7e](https://github.com/gravitational/teleport/pull/40859/commits/ca49e7e131e15bbe0b826f8c0b36cc08ff870c67)

I also changed the connection setup observer to at least be called after we dial GCP.
And I added message received metrics for client/server.

Reply:
![image](https://github.com/gravitational/teleport/assets/16846283/bae4aa95-08b2-47c2-93d5-b57a715a9fda)

I inserted a million rows in my db's "singers" table and then ran `select * from singers`.

The metrics reflect that the client sent only a few messages whereas we 50+ from the server (it's streaming pages of rows back).

Reply:
Why are we checking against `InPayload`? do we need to check `RPCStats.IsClient`?

Reply:
`InPayload` is an incoming payload. `IsClient` indicates whether the stream is a `grpc.ClientStream` or a `grpc.ServerStream`, but either way `stats.InPayload` is recorded when the stream receives a message.

For simplicity, I made the stats handler only concerned with recording any payload it receives.
For our connection to GCP, I set up an instance of this handler to record every message received from GCP.
For our grpc server, I set up an instance of the handler that records every message received from the user.